File: app/services/terrain/artifact.rb
require "digest"
require "time"

module Terrain
  class Artifact
    class Invalid < StandardError; end

    SCHEMA_VERSION = 1
    PROJECTION = {
      "type" => "linear-lat-lng-bounds",
      "x_axis" => "longitude-west-to-east",
      "z_axis" => "latitude-north-to-south",
      "elevation_unit" => "feet"
    }.freeze
    ISO8601 = /\A\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})\z/

    attr_reader :data, :sha256

    def self.write(payload, to:)
      data = JSON.parse(JSON.generate(payload)).merge(
        "schema_version" => SCHEMA_VERSION,
        "projection" => PROJECTION
      )
      new(data)
      bytes = "#{JSON.pretty_generate(data)}\n"

      output_path = Pathname(to)
      FileUtils.mkdir_p(output_path.dirname)
      File.binwrite(output_path, bytes)
      new(data, bytes: bytes)
    end

    def self.read(reference, from:, race_slug:)
      validate_reference!(reference)
      path = Pathname(from)
      invalid_reference!("file is missing") unless path.file?

      bytes = File.binread(path)
      sha256 = Digest::SHA256.hexdigest(bytes)
      invalid_reference!("SHA-256 digest does not match") unless sha256 == reference.fetch("sha256")

      artifact = new(JSON.parse(bytes), bytes: bytes)
      invalid_reference!("Race slug does not match") unless artifact.data.dig("race", "slug") == race_slug
      artifact
    rescue JSON::ParserError => error
      raise Invalid, "Invalid Terrain artifact: JSON could not be parsed (#{error.message})"
    end

    def self.runtime_reference(reference)
      validate_reference!(reference)
      separator = reference.fetch("path").include?("?") ? "&" : "?"
      reference.merge("path" => "#{reference.fetch('path')}#{separator}v=#{reference.fetch('sha256')}")
    end

    def self.validate_reference!(reference)
      invalid_reference!("reference is missing") unless reference.is_a?(Hash)
      invalid_reference!("status is not generated") unless reference["status"] == "generated"
      invalid_reference!("path must identify an owned Terrain artifact") unless reference["path"].is_a?(String) && reference["path"].start_with?("/terrain/")
      invalid_reference!("schema version is unsupported") unless reference["schema_version"] == SCHEMA_VERSION
      invalid_reference!("projection is unsupported") unless reference["projection"] == PROJECTION.fetch("type")
      invalid_reference!("SHA-256 digest is invalid") unless reference["sha256"].is_a?(String) && reference["sha256"].match?(/\A[0-9a-f]{64}\z/)
    end

    def self.invalid_reference!(message)
      raise Invalid, "Invalid Terrain artifact reference: #{message}"
    end
    private_class_method :validate_reference!, :invalid_reference!

    def initialize(data, bytes: nil)
      @data = deep_freeze(data)
      validate!
      @sha256 = Digest::SHA256.hexdigest(bytes || JSON.generate(data))
    end

    def reference(path:)
      {
        "status" => "generated",
        "path" => path,
        "schema_version" => SCHEMA_VERSION,
        "sha256" => sha256,
        "projection" => PROJECTION.fetch("type")
      }
    end

    private

    def deep_freeze(value)
      case value
      when Hash
        value.each { |key, nested| deep_freeze(key); deep_freeze(nested) }
      when Array
        value.each { |nested| deep_freeze(nested) }
      end
      value.freeze
    end

    def validate!
      invalid!("schema version is unsupported") unless data["schema_version"] == SCHEMA_VERSION
      invalid!("projection is unsupported") unless data["projection"] == PROJECTION

      race = data["race"]
      invalid!("race metadata is missing") unless race.is_a?(Hash)
      invalid!("race slug is missing") if race["slug"].blank?
      invalid!("race name is missing") if race["name"].blank?
      invalid!("race year is missing") unless race["year"].is_a?(Integer)
      validate_generated_at!

      source = data["source"]
      invalid!("source metadata is missing") unless source.is_a?(Hash)
      %w[label url attribution].each do |field|
        invalid!("source #{field} is missing") if source[field].blank?
      end

      grid = data["grid"]
      invalid!("grid is missing") unless grid.is_a?(Hash)
      size = grid["size"]
      invalid!("grid size must be at least 2") unless size.is_a?(Integer) && size >= 2
      invalid!("grid zoom is missing") unless grid["zoom"].is_a?(Integer)
      invalid!("grid elevations do not match its size") unless grid["elevations_ft"].is_a?(Array) && grid["elevations_ft"].size == size * size
      invalid!("grid elevations must be numeric") unless grid["elevations_ft"].all? { |value| finite_number?(value) }
      invalid!("grid elevation range is invalid") unless finite_number?(grid["min_ft"]) && finite_number?(grid["max_ft"]) && grid["min_ft"] <= grid["max_ft"]

      bounds = grid["bounds"]
      invalid!("grid bounds are missing") unless bounds.is_a?(Hash)
      %w[min_lat max_lat min_lng max_lng].each do |field|
        invalid!("grid bound #{field} is invalid") unless finite_number?(bounds[field])
      end
      invalid!("grid latitude bounds are invalid") unless bounds["min_lat"] < bounds["max_lat"]
      invalid!("grid longitude bounds are invalid") unless bounds["min_lng"] < bounds["max_lng"]

      profile = data["course_grade_profile"]
      invalid!("course grade profile is missing") unless profile.is_a?(Hash) && profile["segments"].is_a?(Array)
      profile["segments"].each do |segment|
        invalid!("course grade segment is invalid") unless segment.is_a?(Hash)
        %w[from to].each do |endpoint|
          coordinates = segment[endpoint]
          unless coordinates.is_a?(Array) && coordinates.size == 2 && coordinates.all? { |value| finite_number?(value) }
            invalid!("course grade segment #{endpoint} coordinates are invalid")
          end
        end
        invalid!("course grade segment grade is invalid") unless finite_number?(segment["grade_pct"])
        invalid!("course grade segment steepness is invalid") unless segment["steepness"].in?(%w[flat moderate steep])
      end
    end

    def validate_generated_at!
      generated_at = data["generated_at"]
      invalid!("generated at metadata is missing") unless generated_at.is_a?(String) && generated_at.match?(ISO8601)
      Time.iso8601(generated_at)
    rescue ArgumentError
      invalid!("generated at metadata is invalid")
    end

    def finite_number?(value)
      value.is_a?(Numeric) && value.finite?
    end

    def invalid!(message)
      raise Invalid, "Invalid Terrain artifact: #{message}"
    end
  end
end

File: app/javascript/terrain/artifact.js
const SCHEMA_VERSION = 1
const PROJECTION = {
  type: "linear-lat-lng-bounds",
  x_axis: "longitude-west-to-east",
  z_axis: "latitude-north-to-south",
  elevation_unit: "feet"
}
const GENERATED_AT = /^(?<year>\d{4})-(?<month>0[1-9]|1[0-2])-(?<day>0[1-9]|[12]\d|3[01])T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\dZ$/

export async function loadTerrainArtifact(reference, { raceSlug }) {
  validateReference(reference)

  const response = await fetch(reference.path)
  if (!response.ok) throw new Error(`Terrain artifact request failed: HTTP ${response.status}`)

  const bytes = new Uint8Array(await response.arrayBuffer())
  const digest = await sha256(bytes)
  if (digest !== reference.sha256) throw new Error("Terrain artifact SHA-256 digest does not match")

  let artifact
  try {
    artifact = JSON.parse(new TextDecoder().decode(bytes))
  } catch (_error) {
    throw new Error("Terrain artifact JSON is invalid")
  }

  validateArtifact(artifact, { raceSlug })
  return artifact
}

function validateReference(reference) {
  if (!reference || reference.status !== "generated") invalid("reference status is invalid")
  if (typeof reference.path !== "string" || !reference.path.startsWith("/terrain/")) invalid("reference path is invalid")
  if (reference.schema_version !== SCHEMA_VERSION) invalid("reference schema version is unsupported")
  if (reference.projection !== PROJECTION.type) invalid("reference projection is unsupported")
  if (typeof reference.sha256 !== "string" || !/^[0-9a-f]{64}$/.test(reference.sha256)) invalid("reference SHA-256 digest is invalid")
}

function validateArtifact(artifact, { raceSlug }) {
  if (!artifact || artifact.schema_version !== SCHEMA_VERSION) invalid("schema version is unsupported")
  if (!sameProjection(artifact.projection)) invalid("projection is unsupported")
  if (artifact.race?.slug !== raceSlug) invalid("Race slug does not match")
  if (blank(artifact.race?.name)) invalid("Race name is missing")
  if (!Number.isInteger(artifact.race?.year)) invalid("Race year is missing")
  if (!validGeneratedAt(artifact.generated_at)) invalid("generated at metadata is invalid")
  if ([artifact.source?.label, artifact.source?.url, artifact.source?.attribution].some(blank)) invalid("source metadata is incomplete")

  const grid = artifact.grid
  if (!grid || !Number.isInteger(grid.size) || grid.size < 2) invalid("grid size is invalid")
  if (!Number.isInteger(grid.zoom)) invalid("grid zoom is missing")
  if (!Array.isArray(grid.elevations_ft) || grid.elevations_ft.length !== grid.size * grid.size) invalid("grid elevations do not match its size")
  if (!grid.elevations_ft.every(Number.isFinite)) invalid("grid elevations must be numeric")
  if (!Number.isFinite(grid.min_ft) || !Number.isFinite(grid.max_ft) || grid.min_ft > grid.max_ft) invalid("grid elevation range is invalid")

  const bounds = grid.bounds
  if (!bounds || ![bounds.min_lat, bounds.max_lat, bounds.min_lng, bounds.max_lng].every(Number.isFinite)) invalid("grid bounds are invalid")
  if (bounds.min_lat >= bounds.max_lat || bounds.min_lng >= bounds.max_lng) invalid("grid bounds are invalid")
  const segments = artifact.course_grade_profile?.segments
  if (!Array.isArray(segments)) invalid("course grade profile is invalid")
  for (const segment of segments) {
    if (!segment || !coordinatePair(segment.from) || !coordinatePair(segment.to)) invalid("course grade segment coordinates are invalid")
    if (!Number.isFinite(segment.grade_pct)) invalid("course grade segment grade is invalid")
    if (!["flat", "moderate", "steep"].includes(segment.steepness)) invalid("course grade segment steepness is invalid")
  }
}

function coordinatePair(value) {
  return Array.isArray(value) && value.length === 2 && value.every(Number.isFinite)
}

function blank(value) {
  return typeof value !== "string" || value.trim().length === 0
}

function validGeneratedAt(value) {
  const match = typeof value === "string" && value.match(GENERATED_AT)
  if (!match) return false

  const year = Number(match.groups.year)
  const month = Number(match.groups.month)
  const day = Number(match.groups.day)
  const daysInMonth = [31, leapYear(year) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
  return day <= daysInMonth[month - 1]
}

function leapYear(year) {
  return year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0)
}

function sameProjection(projection) {
  return projection && Object.keys(projection).length === Object.keys(PROJECTION).length &&
    projection.type === PROJECTION.type &&
    projection.x_axis === PROJECTION.x_axis &&
    projection.z_axis === PROJECTION.z_axis &&
    projection.elevation_unit === PROJECTION.elevation_unit
}

async function sha256(bytes) {
  if (!globalThis.crypto?.subtle) throw new Error("Terrain artifact SHA-256 validation is unavailable")

  const digest = await globalThis.crypto.subtle.digest("SHA-256", bytes)
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, "0")).join("")
}

function invalid(message) {
  throw new Error(`Invalid Terrain artifact: ${message}`)
}
